### ISSUE TITLE
Is it possible to configure Archive using Amazon S3? (OHIF Viewer)

### ISSUE THREAD

**Issue by johannamussolini** [OHIF/Viewers#355]
Hi all,
Do you know if it's possible to configure the OHIF Viewer or the Standalone Viewer to use Amazon S3 as the Archive instead of Orthanc? does somebody already did it?

I was working with the Standalone Viewer and for the folder containing the studies I was able to create the script which parses it and create the corresponding JSON file, however... I'm not sure how to make Amazon S3 to provide me URLs for each file I've uploaded (so the JSON can have them inside for each file) ... probably should have to make them as public and that's not possible in my case... Should I have to use the OHIF Viewer and not the Standalone Viewer... and just configure my Amazon S3 server as the OHIF archive server? Anyway I'm not sure how my S3 will provide authorization to OHIF app to see the files...

Am I doing something wrong?

Any help would be really appreciated. best regards!

**Comment by TZubiri** [comment]
At the moment, I am trying to delegate the S3 handling to the PACS server (Orthanc or Dcm4chee), otherwise one would need to develop an s3 library with server-side request signing. I think it's simpler to implement the s3 interface PACS side and leave OHIF with a DICOMWeb interface.
I know both Orthanc and dcm4chee have s3 plugins. Haven't got either of those to work yet.

**Comment by diaztula** [comment]
Yes, it is possible, but you must be careful about security of your DICOM files.

If you put your DICOM files in a private S3 bucket, you will need to get a presigned URL for each single file using the AWS API.  This URL allows you (and anybody who get access to the URL) to download the files via http/https and you can define an expiration time for the URL (so the link is no longer valid after X seconds). 

One problem with this approach is that somebody could look at the developer console in the browser and get the links to download your files. But I think that other approaches (such as using tokens, passwords etc) would have the same security issue (somebody could just get the token and make requests to get the DICOM files from the dicom server). Is you use https in the Standalone Viewer, the links will be encrypted in transit.

**Comment by TZubiri** [comment]
> One problem with this approach is that somebody could look at the developer console in the browser and get the links to download your files.

What would be the problem with this? The user could only get the links to images they were already allowed to view in OHIF.

**Comment by TZubiri** [comment]
> > One problem with this approach is that somebody could look at the developer console in the browser and get the links to download your files.
> 
> What would be the problem with this? The user could only get the links to images they were already allowed to view in OHIF.

Furthermore, this approach is safer than using OHIF default Viewer. Currently, if I give a link of a specific study to somebody, I am also giving them access to the study-list. If they can take a look at the developer console, they can use the provided wado proxy to do whatever queries they please, effectively using the PACS account that was configured at config.json

I wouldn't trust OHIF with security and access control. It's just a viewer.

**Comment by johannamussolini** [comment]
Hi guys! thanks for your prompt responses. So... summarizing:

1. @diaztula Do you suggest to use the Standalone Viewer instead of the OHIF Viewer... and handle the JSON creation in a backend server (which could be AWS Lambda function to resolve security issues and forgot about the server). Am I right? 
2. @TZubiri I'm concerned about using the OHIF Viewer and relay on DCM4CHEE as my backend server.... I'm not sure if I understand very well what the DCM4CHEE server is... but it seems to have its own DB and it's made in J2EE... Is the DCM4CHEE a PAC Server? do you know how it works? I'm a little bit confused about it but it seems that is Open Source... 
2. If I create the JSON using the pre-signed URL from AWS API... then each time that somebody want to see his study will be consuming AWS requests and will be charged.. Is that efficient? Isn't it better to download the study in a backend server and create the JSON in such way that the URLs are not pointing to S3 directly and the front end just access the files from the backend (and avoid S3 charges, for example if the user plays a video and see the same study several times...)... 
3. Another issue I have to create the JSON using pre-signed URLs from AWS S3 API is that it has of course a timeout so...the JSON file will change every time, so I have to generate it every time the user wants to see his study or the timeout occur, and read the files from S3 to parse and generate the JSON.. which will also consume S3 requests and would be expensive... 

Any advise you could provide would be really appreciated! best regards to both!

**Comment by diaztula** [comment]
> Hi guys! thanks for your prompt responses. So... summarizing:
> 
> 1. @diaztula Do you suggest to use the Standalone Viewer instead of the OHIF Viewer... and handle the JSON creation in a backend server (which could be AWS Lambda function to resolve security issues and forgot about the server). Am I right?
Yes, that's it. 

> 3. If I create the JSON using the pre-signed URL from AWS API... then each time that somebody want to see his study will be consuming AWS requests and will be charged.. Is that efficient? Isn't it better to download the study in a backend server and create the JSON in such way that the URLs are not pointing to S3 directly and the front end just access the files from the backend (and avoid S3 charges, for example if the user plays a video and see the same study several times...)...

Yes, if you want to keep your files safer you would generate a presigned URL every time the study is opened. 

For large studies (e.g. CTs with thousands of files) and/or many clients using the viewer at the same time, performance could be an issue, since for each file you'll need to call presign from AWS and there is not batch presign AFAIK. 

If you download your files from S3 to a server *outside* AWS (or inside AWS but in a different region) then you will pay download transfer anyway. And you will have to serve those files from your server so the OHIF can read them.

> 4. Another issue I have to create the JSON using pre-signed URLs from AWS S3 API is that it has of course a timeout so...the JSON file will change every time, so I have to generate it every time the user wants to see his study or the timeout occur, and read the files from S3 to parse and generate the JSON.. which will also consume S3 requests and would be expensive...

You could create presigned links that last at most one year (https://github.com/aws/aws-sdk-net/issues/528), or make the files publicly available inside your bucket, but this would be a security issue. 

> 
> Any advise you could provide would be really appreciated! best regards to both!

Well, depending on the volume of data you expect to handle, it may worth using S3 as a storage for your DICOM. It will depend on your needs.

**Comment by TZubiri** [comment]
2- DCM4CHEE or Orthanc are PACS servers, the Non standalone viewer requires a PACS server, this interaction is defined in the DICOM standard, both OHIF and PACS servers provide the necessary information about how to connect in their documentation. It is a cumbersome dependency, but you might reap the benefits and reuse it later, for example, when you need to receive images from medical equipment (which are designed to communicate with PACS servers.)
3- This is an interesting point, what you are describing is a [cache](https://en.wikipedia.org/wiki/Cache_(computing)). There are three nodes in the design, S3, the server and the client. You might choose to cache the images in the server or in the client if you wish to save some money. You will be left to implement this cache on your own. I haven't checked whether the s3 plugins for Orthanc or DCM4Chee provide caching. But it would be a nice place to put this functionality, in order to avoid having devs like us develop caches multiple times. As diaztula mentions, since S3 and the server will be in the same region, this isn't an issue. This would only make sense for applications that are distributed globally. I don't know your use case, but it's pretty rare in healthcare.
4- Shouldn't be a problem, the presigning of requests should be automated.

**Comment by johannamussolini** [comment]
> and there is not batch presign AFAIK.

thanks @diaztula, as you say everything depends on our needs and the volume of data. Currently the volume of data is huge so that's why we are looking for cloud storage archive (like S3) so forgot about space and costs related to generating backups, etc.

1- Thanks for your advise. I also think that using the standalone viewer is better (the only disadvantage is that the tool bar cannot be changed you have to use it as it is.. and the other one is that you have to handle the JSON creation and handle very well the backend server side, I've never used AWS Lambda functions yet, but I'll research about them). On the other hand, The OHIF Viewer approach relays the JSON creation (I guess) to the PACS server (i.e. DCM4CHEE has) and I'm not sure if I would

>

**Comment by johannamussolini** [comment]
@diaztula 

> 1. @diaztula Do you suggest to use the Standalone Viewer instead of the OHIF Viewer... and handle the JSON creation in a backend server (which could be AWS Lambda function to resolve security issues and forgot about the server). Am I right?
   **Yes, that's it.**

I also think that using the Standalone Viewer is a better approach than using the OHIF Viewer. The only disadvantage is that in the Standalone you cannot change the tool bar (as far as I know) and you need to handle the JSON creation and be careful when coding the backend server for that (or use AWS Lambda functions which I've never used before but I'll research about them), maybe using a Lambda function in the same region as S3 has no charge when reading files to S3 to get the pre-signed URLs for each DICOM file...

> Yes, if you want to keep your files safer you would generate a presigned URL every time the study is opened.

that's the idea.... to keep it safer

> For large studies (e.g. CTs with thousands of files) and/or many clients using the viewer at the same time, performance could be an issue, since for each file you'll need to call presign from AWS and there is not batch presign AFAIK.

that's my problem :(

> If you download your files from S3 to a server _outside_ AWS (or inside AWS but in a different region) then you will pay download transfer anyway. And you will have to serve those files from your server so the OHIF can read them.

yes but I was thinking that I'll just pay once that is when you download the study and then you can read it multiple times with no charge from S3 (I'm still not sure if S3 charges for reading a file... instead of downloading it). Anyway this approach has also an issue: I'll have to implement a kind of CACHE (small filesystem) in my backend server and take care about it's security... and that could take a lot of time and I'm not an expert on that matter....

On the other hand, I was also thinking that maybe it's possible to  **_create the JSON just one time_** and keep it in S3 with the study... and then when an user needs to see his study then in some way parse it and just update the URLs inside with the new pre-signed URLs and then reduce the cost to read each file again in S3 to generate the JSON... Not sure if you understand the idea.

> You could create presigned links that last at most one year ([aws/aws-sdk-net#528](https://github.com/aws/aws-sdk-net/issues/528)), or make the files publicly available inside your bucket, but this would be a security issue.

Yes I know, but as you said before that would not be secure, the idea is to create them with an hour of timeout for example, and then if the user wants to see his study again the pre-signed URLs will be generated again, but again not sure for large studies how long could take or if could be a performance issue,..

> Well, depending on the volume of data you expect to handle, it may worth using S3 as a storage for your DICOM. It will depend on your needs.

Yes, that's for sure completely right. The volume of data is huge (need to handle a lot of different medical images, not just CTs) and I think S3 is one of the best solutions (since they will pay for the space they consume and most important thing: will never loose the data). But since I'm pretty new working with DICOM images and PACS servers, etc... I was just looking for some advise from people who has more knowledge and can provide some guidance, or just say me "hey! you are misunderstanding this!... " :P LOL 

Again, thanks so much for spending your time on replying back to me! I really appreciate any help from the experts.

**Comment by johannamussolini** [comment]
@TZubiri 👍 

> 2- DCM4CHEE or Orthanc are PACS servers, the Non standalone viewer requires a PACS server, this interaction is defined in the DICOM standard, both OHIF and PACS servers provide the necessary information about how to connect in their documentation. It is a cumbersome dependency, but you might reap the benefits and reuse it later, for example, when you need to receive images from medical equipment (which are designed to communicate with PACS servers.)

Yes that's what I understood. However the user already have a PACS server (but the host it locally... with K-PAC... as far as I was able to analyze... it's not a great solution and they have lost several times studies because they are storing them locally... crazy thing)
So, when looking at Orthanc (I'm really not sure if I could use it for my solution since it's not completely free... unless I understood wrong). DCM4CHEE seems to be Open Source but (like Orthanc) you need to relay on it's own Architecture (J2EE) it's own DB.. etc... and would be replacing the PACS Server that my user already has installed (K-PAC) and that could be a huge change for them.... Also, I'm not sure about which kind of AWS Server should I have to pick to host this kind of server... On the other hand it seems it's resolves the JSON creation and has the ONLINE and NEARLINE Storage... but not sure if the ONLINE is inside the same server (so it could serve as the CACHE I'd have to develop if I use the Standalone Viewer) and the NEARLINE is S3... or how it works,.. It seems that DCM4CHEE servers resolves a lot of things... So I'm not sure if it would save me a lot of time and head aches, but again I'll relay on their architecture... and currently my backend is Node.js express server... which is something completely different to J2EE,.... and not sure is a good design solution to mix them... or use two different server... I'll have to research deeper about what's the best design and solution.

> 3- This is an interesting point, what you are describing is a [cache](https://en.wikipedia.org/wiki/Cache_(computing)). There are three nodes in the design, S3, the server and the client. You might choose to cache the images in the server or in the client if you wish to save some money. You will be left to implement this cache on your own. I haven't checked whether the s3 plugins for Orthanc or DCM4Chee provide caching. But it would be a nice place to put this functionality, in order to avoid having devs like us develop caches multiple times. **As diaztula mentions, since S3 and the server will be in the same region, this isn't an issue**. This would only make sense for applications that are distributed globally. I don't know your use case, but it's pretty rare in healthcare

So, you guys say that if S3 and the backend server are both in the same AWS region then the Cache will not be necessary? I didn't get that. could you please clarify?

> 4- Shouldn't be a problem, the presigning of requests should be automated.

No the only problem I see is that for the Standalone Viewer the JSON which needs to be created would need to be generated every time the pre-signed URL expires and that would mead to go to S3, read the entire study again (not sure if S3 charges for that) and generate the JSON with the new S3 URL... As I was saying to @diaztula before,.. maybe I can just generate the JSON once... and then each time the users wants to see the study again, if the URL already expired... then just get the new URLs and in some way insert them in the JSON which already was created... and in that case I should not have to go to S3 again to read each DCM file to extract its data and generate the JSON... I mean the JSON once created from the Study folder (which contains all the DCM files) is always the same... the only thing that will be changing is the URL for each DCM instance... and that the only thing I'd need to update... I don't know... I think it could be another solution to reduce S3 Costs and also improve performance. The JSON can be created when uploading the Study for the first time in S3 and that's all you will not have to read and parse every DICOM file again to get it... just change it to update the URL...

Anyway... Thanks so much for your help and time on replying back to me! 👍

**Comment by TZubiri** [comment]
2- The clinic I am working on also has a local PACS instance, however I am setting up my own Orthanc PACS in parallel, to allow me to have control over images transfered from their existing PACS, or medical equipments. Orthanc was designed to be a complement to traditional PACS servers, so nothing crazy is going on here, I view it as an interoperability tool.
Again, in my project I am trying to delegate the S3 Handling to Orthanc, so Ohif only has to worry about Dicomweb interfaces. This seems like the most reusable solution; by having solutions go through a centralizing technology, I avoid the problem of needing all technologies to interface with each other, which would require a . I don't need an s3 to viewer interface, I need an s3 to PACS technology, and a viewer to PACS technology.  But these 2 technologies will be more popular (and therefore more robust). 
P.S: The amount of interfaces needed for a group of X technologies to have interfaces between all of them would be [C(X,2)](https://www.wolframalpha.com/input/?i=C(X,2))

>So, you guys say that if S3 and the backend server are both in the same AWS region then the Cache will not be necessary? I didn't get that. could you please clarify?
3- Correct.

4- http://wiki.c2.com/?PrematureOptimization I'd say don't worry about it yet, implement it first, then worry about it if it's an issue.

**Comment by johannamussolini** [comment]
@TZubiri thanks so much for your advice. I'll try to research a little bit more until deciding what architecture design to choose, but what you say has sense... reuse OHIF Viewer and Orthanc will save a lot of time. BTW, just last concerns:
 1. Is Orthanc totally free? There is no issue to use it for my application design that is also for a clinic? 
2. Why did you choose Orthanc instead of DCM4CHEE? Is it better option?
Thanks!
